agent/src/main/java/com/teamscale/jacoco/agent/upload/teamscale/DelayedTeamscaleMultiProjectUploader.java

```diff
@@ -20,6 +21,34 @@ public DelayedTeamscaleMultiProjectUploader(BiFunction<String, CommitInfo, IUplo
 		this.uploaderFactory = uploaderFactory;
 	}
 
+	/**
+	 * Checks if there is already a uploader present with the given project and commit. This is relevant to prevent
+	 * uploading coverage multiple times to the same location.
+	 */
+	public boolean projectAndCommitAlreadyRegistered(ProjectAndCommit projectAndCommit) {
+		return teamscaleUploaders.stream().anyMatch(uploader -> {
+			// Might happen during testing.
+			if (uploader == null) {
+				return false;
+			}
+			TeamscaleUploader teamscaleUploader = (TeamscaleUploader) uploader;
+			TeamscaleServer teamscaleServer = teamscaleUploader.getTeamscaleServer();
+			if (!teamscaleServer.project.equals(projectAndCommit.getProject())) {
+				return false;
+			}
+
+			CommitInfo newCommitInfo = projectAndCommit.getCommitInfo();
+			if (newCommitInfo.preferCommitDescriptorOverRevision) {
+				return newCommitInfo.commit.equals(teamscaleServer.commit);
+			}
+			if (newCommitInfo.revision != null) {
+				return newCommitInfo.revision.equals(teamscaleServer.revision);
+			}
+			// Needs to be checked a second time, as it can be that just git.commit.time and git.branch is set,
+			// which does not set CommitInfo#preferCommitDescriptorOverRevision.
+			return newCommitInfo.commit.equals(teamscaleServer.commit);
```

Comment:
We are duplicating some of the domain knowledge (i.e. when to pick which commit vs revision) from `AgentOptions#createTeamscaleMultiProjectUploader`, which means we need to keep both places in sync.
I would suggest to:
- Repurpose the "`uploaderFactory`" to a `teamscaleServerFactory` that would just create the `TeamscaleServer` object
- This can be called in `setTeamscaleProjectForRevision`
- Then you can check whether the new `TeamscaleServer` object matches any of the existing `TeamscaleServer` objects (i.e. implement `TeamscaleServer#hasSameProjectAndCommit(TeamscaleServer)` method or similar) and if not create and add the new uploader
- To keep you test working you can add a second `Function<TeamscaleServer, IUploader>` for the actual uploader creation

Reply:
Great suggestions. I decided against adding a new Function<TeamscaleServer, IUploader> for testing purposes but instead added a getter for the teamscaleUploaders. In addition I restructured the test a little bit. Let me know what you think :)